File: apps/api/src/hosts/conn.ts
// One connected host: multiplexes PTY and preview-port streams over its single
// WebSocket using the binary framing. Knows nothing about enrollment or the store.
import { Frame } from '@sandboxd/core/framing'
import type { Msg } from '@sandboxd/core/messages'

export interface StreamSub {
  onData(d: Uint8Array): void
  onReplay?(): void
  onOpen?(): void
  onError?(msg: string): void
  onClose(): void
}

export interface PtyHandle {
  write(d: Uint8Array): void
  resize(size: Msg.Size): void
  close(): void
}
export interface PortHandle {
  write(d: Uint8Array): void
  close(): void
}
export interface Capacity {
  running: number
  max: number
}

/** What a HostConn needs from its socket. Bun's ServerWebSocket satisfies it; tests pass a stub. */
export interface Transport {
  send(data: string | Uint8Array): unknown
  close(): void
}

/** Host messages that address one stream rather than the host as a whole. */
export type StreamMsg = Extract<Msg.Host, { stream: number }>

export class HostConn {
  private streams = new Map<number, StreamSub>()
  /** CP-allocated stream ids are odd and never reused within a connection. */
  private nextStream = 1
  running: number
  max: number

  constructor(
    readonly hostId: string,
    readonly transport: Transport,
    running: number,
    max: number,
  ) {
    this.running = running
    this.max = max
  }

  capacity(): Capacity {
    return { running: this.running, max: this.max }
  }
  send(msg: Msg.Cp) {
    this.transport.send(JSON.stringify(msg))
  }
  close() {
    this.transport.close()
  }

  // ---- inbound from the host ----
  onFrame(stream: number, payload: Uint8Array) {
    this.streams.get(stream)?.onData(payload)
  }

  onStreamMsg(msg: StreamMsg) {
    switch (msg.type) {
      case 'pty.replay':
        this.streams.get(msg.stream)?.onReplay?.()
        return
      case 'port.open':
        this.streams.get(msg.stream)?.onOpen?.()
        return
      case 'port.error':
        this.streams.get(msg.stream)?.onError?.(msg.msg)
        this.endStream(msg.stream)
        return
      case 'pty.closed':
      case 'port.close':
        this.endStream(msg.stream)
        return
    }
  }

  /** The socket is gone: every subscriber learns its stream is closed. */
  closeAll() {
    const subs = [...this.streams.values()]
    this.streams.clear()
    for (const s of subs) s.onClose()
  }

  private endStream(stream: number) {
    const s = this.streams.get(stream)
    if (!s) return
    this.streams.delete(stream)
    s.onClose()
  }

  private allocStream(sub: StreamSub): number {
    const stream = this.nextStream
    this.nextStream += 2
    this.streams.set(stream, sub)
    return stream
  }

  // ---- commands to the host ----
  createSession(spec: Msg.Spec) {
    this.send({ type: 'session.create', spec })
    this.running += 1 // optimistic until the next heartbeat
  }

  destroySession(sid: string) {
    this.send({ type: 'session.destroy', sid })
  }

  openPty(sid: string, size: Msg.Size, sub: StreamSub): PtyHandle {
    const stream = this.allocStream(sub)
    this.send({ type: 'pty.open', sid, stream, size })
    return {
      write: (d) => this.transport.send(Frame.encode(stream, d)),
      resize: (sz) => this.send({ type: 'pty.resize', sid, size: sz }),
      close: () => {
        if (this.streams.delete(stream)) this.send({ type: 'pty.close', stream })
      },
    }
  }

  /** Resolves on `port.open`, rejects on `port.error`. `sub.onClose` fires in both the error and the normal-close case. */
  dialPort(
    sid: string,
    port: number,
    sub: Pick<StreamSub, 'onData' | 'onClose'>,
  ): Promise<PortHandle> {
    const { promise, resolve, reject } = Promise.withResolvers<PortHandle>()
    let stream = 0
    const handle: PortHandle = {
      write: (d) => this.transport.send(Frame.encode(stream, d)),
      close: () => {
        if (this.streams.delete(stream)) this.send({ type: 'port.close', stream })
      },
    }
    stream = this.allocStream({
      onData: sub.onData,
      onOpen: () => resolve(handle),
      onError: (m) => reject(new Error(m)),
      onClose: sub.onClose,
    })
    this.send({ type: 'port.dial', sid, port, stream })
    return promise
  }
}

File: internal/cp/hosts/conn.go
// Package hosts owns the worker tunnels: one WebSocket per host, multiplexed into
// streams, plus enrollment and the host admin use cases. Everything above it — the
// scheduler, the attach bridge, the preview proxy — sees either a narrow interface or a
// net.Conn, never the socket.
package hosts

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"net"
	"os"
	"sync"
	"sync/atomic"
	"time"

	"github.com/coder/websocket"

	"sandboxd/internal/cp"
	"sandboxd/internal/wire"
)

const (
	// writeQueue is the buffer in front of the one writer goroutine.
	writeQueue = 256
	// writeTimeout bounds a single frame. A socket that cannot take a frame in this long
	// is wedged, and dropping it beats stalling every stream behind it.
	writeTimeout = 10 * time.Second
	// controlWait bounds how long a control message waits for room on the queue. See
	// send: the caller is usually the scheduler goroutine, and it may not wait longer.
	controlWait = time.Second
	readLimit   = 4 << 20
)

var errHostGone = errors.New("hosts: host is offline")

type outgoing struct {
	typ  websocket.MessageType
	data []byte
	// ack, when set, is closed once this entry has left the queue. An entry with no data
	// is an ack and nothing else, which is how flush waits for the socket to catch up.
	ack chan struct{}
}

// Conn is one connected host. It knows nothing about enrollment or the store: the hub
// decides who this is, and the conn multiplexes.
type Conn struct {
	ws  *websocket.Conn
	log *slog.Logger

	// hostID is the host this socket enrolled as. It is written once, on the read
	// goroutine, but writeLoop is already running by then and names the host when a write
	// fails, so it is atomic rather than plain.
	hostID atomic.Pointer[string]

	out       chan outgoing
	done      chan struct{}
	closeOnce sync.Once

	mu      sync.Mutex
	running int
	max     int
	// next is the id of the stream to allocate. CP-allocated ids are odd and are never
	// reused within a connection (DESIGN.md, "Wire protocol").
	next    uint32
	streams map[uint32]*Stream
}

func newConn(ws *websocket.Conn, log *slog.Logger) *Conn {
	return &Conn{
		ws:      ws,
		log:     log,
		out:     make(chan outgoing, writeQueue),
		done:    make(chan struct{}),
		next:    1,
		streams: map[uint32]*Stream{},
	}
}

// HostID is the host this socket enrolled as, empty until it has said hello.
func (c *Conn) HostID() string {
	if id := c.hostID.Load(); id != nil {
		return *id
	}
	return ""
}

// setHostID names the socket, once, when enrollment has decided who it is.
func (c *Conn) setHostID(id string) { c.hostID.Store(&id) }

// Capacity is what the host last reported.
func (c *Conn) Capacity() cp.Capacity {
	c.mu.Lock()
	defer c.mu.Unlock()
	return cp.Capacity{Running: c.running, Max: c.max}
}

func (c *Conn) setCapacity(running, max int) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.running, c.max = running, max
}

// --- the one writer -------------------------------------------------------------------

// writeLoop is the only thing that calls Write on the socket. It ends when the queue's
// owner shuts the connection down, when the server stops, or on the first write error.
func (c *Conn) writeLoop(ctx context.Context) {
	defer c.shutdown()
	for {
		select {
		case <-ctx.Done():
			return
		case <-c.done:
			return
		case m := <-c.out:
			if m.data != nil {
				wctx, cancel := context.WithTimeout(ctx, writeTimeout)
				err := c.ws.Write(wctx, m.typ, m.data)
				cancel()
				if err != nil {
					c.log.Warn("tunnel write failed", "host_id", c.HostID(), "err", err)
					return
				}
			}
			if m.ack != nil {
				close(m.ack)
			}
		}
	}
}

// flush waits for everything already queued to reach the socket, so a message sent just
// before the connection is dropped — hello.rejected — is not lost to the close.
func (c *Conn) flush(timeout time.Duration) {
	ack := make(chan struct{})
	select {
	case c.out <- outgoing{ack: ack}:
	case <-c.done:
		return
	}
	select {
	case <-ack:
	case <-c.done:
	case <-time.After(timeout):
	}
}

// send queues a control message. Control is never dropped silently: a lost sandbox.create
// would leave a row in `creating` with nothing behind it, so a caller that cannot queue
// one is told the host is gone and treats it as an unplaced sandbox.
//
// The policy when the queue is full, which this stream kind owes the same way a PTY and a
// port stream do: wait controlWait, then drop the tunnel. The caller here is usually the
// scheduler goroutine — the only writer of sandbox state — and the queue it is waiting on
// carries every PTY and port frame for this host too. Blocking until writeTimeout retired
// the socket would stall every API call in the process behind one wedged worker. A host
// that cannot take a control frame in a second is gone; dropping it makes the worker
// reconnect, and hello reconciles both directions.
func (c *Conn) send(m wire.FromCP) error {
	b, err := wire.Marshal(m)
	if err != nil {
		return err
	}
	timer := time.NewTimer(controlWait)
	defer timer.Stop()
	select {
	case c.out <- outgoing{typ: websocket.MessageText, data: b}:
		return nil
	case <-c.done:
		return errHostGone
	case <-timer.C:
		c.log.Warn("control queue stalled; dropping the tunnel",
			"host_id", c.HostID(), "after", controlWait)
		c.shutdown()
		return errHostGone
	}
}

// sendFrame queues stream bytes, honouring the stream's own lifetime and write deadline.
func (c *Conn) sendFrame(id uint32, payload []byte, done, deadline <-chan struct{}) error {
	select {
	case <-done:
		return net.ErrClosed
	default:
	}
	select {
	case c.out <- outgoing{typ: websocket.MessageBinary, data: wire.Encode(id, payload)}:
		return nil
	case <-done:
		return net.ErrClosed
	case <-c.done:
		return errHostGone
	case <-deadline:
		return os.ErrDeadlineExceeded
	}
}

// shutdown ends the socket. Streams are ended by the read goroutine's closeAll, which is
// the only place allowed to close a stream's inbound channel.
func (c *Conn) shutdown() {
	c.closeOnce.Do(func() {
		close(c.done)
		_ = c.ws.CloseNow()
	})
}

// --- streams ---------------------------------------------------------------------------

func (c *Conn) alloc(sid string, port int, pty bool) (*Stream, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	select {
	case <-c.done:
		return nil, errHostGone
	default:
	}
	s := newStream(c, c.next, sid, port, pty)
	c.next += 2
	c.streams[s.id] = s
	return s, nil
}

func (c *Conn) forget(id uint32) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.streams, id)
}

func (c *Conn) stream(id uint32) *Stream {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.streams[id]
}

// openPTY attaches a viewer to a sandbox's terminal. There is no acknowledgement to wait
// for: the host answers with pty.replay and then bytes.
func (c *Conn) openPTY(sid string, size wire.Size) (*Stream, error) {
	s, err := c.alloc(sid, 0, true)
	if err != nil {
		return nil, err
	}
	if err := c.send(&wire.PtyOpen{SID: sid, Stream: s.id, Size: size}); err != nil {
		c.forget(s.id)
		return nil, err
	}
	return s, nil
}

// dial opens a TCP connection inside a sandbox and waits for the host to say whether it
// worked, so the caller gets a connection or an error and never a half-open stream.
func (c *Conn) dial(ctx context.Context, sid string, port int) (*Stream, error) {
	s, err := c.alloc(sid, port, false)
	if err != nil {
		return nil, err
	}
	if err := c.send(&wire.PortDial{SID: sid, Port: port, Stream: s.id}); err != nil {
		c.forget(s.id)
		return nil, err
	}
	if err := s.awaitOpen(ctx, c.done); err != nil {
		// Close is a no-op when the host already ended the stream with port.error.
		_ = s.Close()
		return nil, fmt.Errorf("hosts: dial %s:%d: %w", sid, port, err)
	}
	return s, nil
}

// endStream is the host hanging up: pty.closed, port.close, or port.error.
func (c *Conn) endStream(id uint32) {
	c.mu.Lock()
	s := c.streams[id]
	delete(c.streams, id)
	c.mu.Unlock()
	if s != nil {
		s.finish()
	}
}

// closeAll ends every stream because the socket is gone. Only the read goroutine calls it.
func (c *Conn) closeAll() {
	c.mu.Lock()
	streams := make([]*Stream, 0, len(c.streams))
	for _, s := range c.streams {
		streams = append(streams, s)
	}
	clear(c.streams)
	c.mu.Unlock()
	for _, s := range streams {
		s.finish()
	}
}

// --- inbound ---------------------------------------------------------------------------

func (c *Conn) onFrame(id uint32, payload []byte) {
	if s := c.stream(id); s != nil {
		s.deliver(payload)
	}
}

// onStreamMsg handles the host messages that address one stream rather than the host.
func (c *Conn) onStreamMsg(m wire.FromHost) {
	switch m := m.(type) {
	case *wire.PtyReplay:
		// A marker for the viewer, and the CP forwards bytes verbatim: the browser sees
		// the tail and then live output, which is what "replay" means to it.
	case *wire.PortOpen:
		if s := c.stream(m.Stream); s != nil {
			s.opened(nil)
		}
	case *wire.PortError:
		if s := c.stream(m.Stream); s != nil {
			s.opened(errors.New(m.Msg))
		}
		c.endStream(m.Stream)
	case *wire.PtyClosed:
		c.endStream(m.Stream)
	case *wire.PortClose:
		c.endStream(m.Stream)
	default:
		c.log.Warn("unroutable host message", "host_id", c.HostID(), "type", fmt.Sprintf("%T", m))
	}
}
